### ISSUE TITLE
Submodule update and detached HEAD

### ISSUE THREAD

**Issue by batbaille** [ForkIssues/TrackerWin#569]
Hello, we are a team that works with a repository which contains three submodules.
In our use case, a first developer commits and pushes changes into one submodule. A second developer pulls the changes in the parent repository and updates the submodule (checking the option on pull) and he ends up with a detached HEAD. 
If developper 2 is not careful, it might commit his new changes outside the branch.

I understand that some developpers would like to ignore the submodules mechanic because they don't work with them but the other developers who work with them must be careful to reattach the HEAD to the local branch.

Could we have a option to checkout automatically the local branch after a submodule update ?

**Comment by felixhirt** [comment]
Possible duplicate of #432 
I think what you are describing is the way git submodules work (they do not track branches but commits, and checking out a commit always leads to a detached head). Interesting read: https://codingkilledthecat.wordpress.com/2012/04/28/why-your-company-shouldnt-use-git-submodules/
(We use submodules as well, but it does not hurt to be aware of their shortcomings)

**Comment by StephenHodgson** [comment]
I don't usually have this problem with SourceTree, I wonder what that git client is doing differently.

It's a bit tedious to have to make sure my submodules are checked out correctly. I've accidentally pushed to a detached head quite a few times because of this, but luckily I know how to recover this quickly and merge it back into my local branch.

**Comment by felixhirt** [comment]
I agree it would be nice to check out the branch instead of the commit if the submodule points to a commit which happens to be a branch head. What happens if there are multiple branch heads on the commit though...

**Comment by StephenHodgson** [comment]
Good question. We could fallback to the commit checkout instead

**Comment by StephenHodgson** [comment]
```
git submodule foreach -q --recursive 'git checkout $(git config -f $toplevel/.gitmodules submodule.$name.branch)'
```
Seems to do the trick

**Comment by StephenHodgson** [comment]
This seems to work fairly well
```
git submodule foreach -q --recursive 'echo $name && git switch -C $(git name-rev --exclude=refs/tags/\* --name-only $(git rev-parse @{-1}))'
```

**Comment by Stanzilla** [comment]
I have been bitten by that quite a few times
